wagtail/admin/mail.py

```diff
@@ -85,30 +101,214 @@ def send_notification(page_revision_id, notification, excluded_user_id):
         "settings": settings,
     }
 
-    # Send emails
-    sent_count = 0
-    for recipient in email_recipients:
-        try:
-            # update context with this recipient
-            context["user"] = recipient
-
-            # Translate text to the recipient language settings
-            with override(recipient.wagtail_userprofile.get_preferred_language()):
-                # Get email subject and content
-                email_subject = render_to_string(template_subject, context).strip()
-                email_content = render_to_string(template_text, context).strip()
-
-            kwargs = {}
-            if getattr(settings, 'WAGTAILADMIN_NOTIFICATION_USE_HTML', False):
-                kwargs['html_message'] = render_to_string(template_html, context)
-
-            # Send email
-            send_mail(email_subject, email_content, [recipient.email], **kwargs)
-            sent_count += 1
-        except Exception:
-            logger.exception(
-                "Failed to send notification email '%s' to %s",
-                email_subject, recipient.email
-            )
+    connection = get_connection()
+
+    with OpenedConnection(connection) as open_connection:
+
+        # Send emails
+        sent_count = 0
+        for recipient in email_recipients:
+            try:
+                # update context with this recipient
+                context["user"] = recipient
+
+                # Translate text to the recipient language settings
+                with override(recipient.wagtail_userprofile.get_preferred_language()):
+                    # Get email subject and content
+                    email_subject = render_to_string(template_subject, context).strip()
+                    email_content = render_to_string(template_text, context).strip()
+
+                kwargs = {}
+                if getattr(settings, 'WAGTAILADMIN_NOTIFICATION_USE_HTML', False):
+                    kwargs['html_message'] = render_to_string(template_html, context)
+
+                # Send email
+                print(recipient.email)
+                send_mail(email_subject, email_content, [recipient.email], connection=open_connection, **kwargs)
+                sent_count += 1
+            except Exception:
+                logger.exception(
+                    "Failed to send notification email '%s' to %s",
+                    email_subject, recipient.email
+                )
 
     return sent_count == len(email_recipients)
+
+
+class Notifier:
+    """Class for sending email notifications upon events: callable, taking a notifying_instance and a notification (str)
+    and sending email notifications using rendered templates"""
+
+    def __init__(self, valid_classes, valid_notifications):
+        self.valid_classes = valid_classes
+        self.valid_notifications = valid_notifications
+
+    def can_handle_class(self, instance, **kwargs):
```

Comment:
I assume these have been broken out into their own methods for ease of overriding, but that's opening things up to be misused, as `GroupApprovalTaskStateNotifier` is already demonstrating...

Firstly, the method signature is a bit vague and inconsistent - is it accepting a class or an instance? Is it being passed `kwargs` or not? Those details can be tightened up, but then you run into a second problem: as soon as a `Notifier` subclass overrides this to do something more fancy than checking the type of `instance` against a list of classes, the method name is a lie. In the case of `GroupApprovalTaskStateNotifier`, it's fudging both the identity of the `instance` being tested, and the meaning of `valid_classes`, so it's essentially repurposing this one line of code to do something entirely different. This would be so much clearer if the only method available to be overridden was `can_handle`:

    class GroupApprovalTaskStateNotifier(Notifier):
        def can_handle(self, instance, notification, **kwargs):
            return (
                super().can_handle(instance, notification, **kwargs) and 
                isinstance(instance.task.specific, GroupApprovalTask)
            )

(On that note, I'm using `isinstance` here rather than comparing `type(instance)` so that it recognises subclasses too - I expect we ought to do the same in the code here too)

Reply:
Ah yes, good point. I've removed these other methods as you suggested now!